pipeline/data/dataset_importer.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+"""
+Downloads a dataset and runs augmentation if needed
+
+Example:
+    SRC=ru TRG=en python pipeline/data/dataset_importer.py \
+        --type=corpus \
+        --dataset=sacrebleu_aug-mix_wmt19 \
+        --output_prefix=$(pwd)/test_data/augtest
+"""
+
+import argparse
+import os
+import re
+import subprocess
+import sys
+from typing import Iterable, List
+
+from opustrainer.modifiers.surface import TitleCaseModifier, UpperCaseModifier
+from opustrainer.modifiers.typos import TypoModifier
+from opustrainer.types import Modifier
+
+
+class CompositeModifier:
+    def __init__(self, modifiers: List[Modifier]):
+        self._modifiers = modifiers
+
+    def __call__(self, batch: List[str]) -> Iterable[str]:
+        for mod in self._modifiers:
+            batch = list(mod(batch))
+
+        return batch
+
+
+MIX_PROB = 0.1  # 10% will be augmented in the mix
+
+modifier_map = {
+    "aug-typos": TypoModifier(MIX_PROB),
+    "aug-title": TitleCaseModifier(MIX_PROB),
+    "aug-title-strict": TitleCaseModifier(1.0),
+    "aug-upper": UpperCaseModifier(MIX_PROB),
+    "aug-upper-strict": UpperCaseModifier(1.0),
+    "aug-mix": CompositeModifier(
+        [
+            # TODO: enable typos, issue https://github.com/hplt-project/OpusTrainer/issues/40
+            # TypoModifier(NOISE_RATE),
+            TitleCaseModifier(MIX_PROB),
+            UpperCaseModifier(MIX_PROB),
+        ]
+    ),
+}
+
+
+def run_cmd(cmd: List[str]):
+    result = None
+    try:
+        result = subprocess.run(
+            cmd,
+            universal_newlines=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            check=False,
+        )
+        result.check_returncode()
+    except:
+        if result:
+            print(result.stdout)
+        raise
+
+    print(result.stdout)
+
+
+# we plan to use it only for small evaluation datasets
+def augment(output_prefix: str, aug_modifer: str):
+    if aug_modifer not in modifier_map:
+        raise ValueError(f"Invalid modifier {aug_modifer}. Allowed values: {modifier_map.keys()}")
+
+    modifer = modifier_map[aug_modifer]
+
+    # these envs are standard across the pipeline
+    src = os.environ["SRC"]
+    trg = os.environ["TRG"]
+    comp_cmd = os.getenv("COMPRESSION_CMD", "pigz")
+    comp_ext = os.getenv("ARTIFACT_EXT", "gz")
+
+    uncompressed_src = f"{output_prefix}.{src}"
+    uncompressed_trg = f"{output_prefix}.{trg}"
+    compressed_src = f"{output_prefix}.{src}.{comp_ext}"
+    compressed_trg = f"{output_prefix}.{trg}.{comp_ext}"
+
+    if os.path.isfile(uncompressed_src):
+        os.remove(uncompressed_src)
+    if os.path.isfile(uncompressed_trg):
+        os.remove(uncompressed_trg)
+    run_cmd([comp_cmd, "-d", compressed_src])
+    run_cmd([comp_cmd, "-d", compressed_trg])
+
+    with open(uncompressed_src) as f:
```

Comment:
Issue: This command is reading the entire file into memory, processing it, and writing it out. This means the dataset will be loaded completely into memory, rather than stream it.

Suggestion: How about creating an iterator that streams everything, something like this? (I didn't try running this, but just thought through the problem)

```py
class CorpusIter:
    def __init__(self, f_src_in, f_trg_in, f_src_out, f_trg_out):
        self.f_src = f_src
        self.f_trg = f_trg

    def __iter__(self):
        return self

    def __next__(self):
        src_sentence = self.f_src.readline()
        trg_sentence = self.f_trg.readline()
        return f"{src_sentence}\t{trg_sentence}"

with \
    open(uncompressed_src) as f_src_in, \
    open(uncompressed_trg) as f_trg_in, \
    open(uncompressed_src2, "w") as f_src_out, \
    open(uncompressed_trg2, "w") as f_trg_out:

    corpus = CorpusIter(f_src_in, f_trg_in, f_src_out, f_trg_out)

    for line in modifer(corpus):
        (modified_src, modified_trg) = line.split("\t")
        f_src_out.write(modified_src)
        f_src_out.write("\n")
        f_trg_out.write(modified_trg)
        f_trg_out.write("\n")

```

Reply:
We don't need to optimize this because it's used only for small evaluation datasets. We augment training datasets on the fly. I added a comment about it in [line 79](https://github.com/mozilla/firefox-translations-training/pull/219/files#diff-60c5881d16ade61511201a410163753c2ef9c76cc3189d272ffad40b1568ecadR79).

We would definitely need to stream for large datasets.